File: Estudos/EstruturasFlexiveis/PilhaFlexivel.java
package Estudos.EstruturasFlexiveis;

class Celula{
    public int elemento;
    public Celula prox;

    Celula(){
        this.elemento = 0;
        this.prox = null;
    }
    Celula(int x){
        this.elemento = x;
        this.prox = null;
    }
}

class Pilha{
    private Celula topo;

    Pilha(){
        this.topo = null;
    }

    void inserir(int x){
        Celula tmp = new Celula(x);
        if(topo == null){
            topo = tmp;
        }else{
            tmp.prox = topo;
            topo = tmp;
        }
        tmp = null;
    }

    void mostra(){
        for(Celula i = topo; i != null; i = i.prox){
            System.out.print(i.elemento + " ");
        }
        System.out.println();
    }
}

public class PilhaFlexivel {
    public static void main(String args[]){
        Pilha pilha = new Pilha();
        pilha.inserir(1);
        pilha.mostra();
        pilha.inserir(2);
        pilha.mostra();
        pilha.inserir(3);
        pilha.mostra();
        pilha.inserir(4);
        pilha.mostra();

    }
}

File: Estudos/EstruturasFlexiveis/PilhaFlexivel/PilhaFlexivel.c
#include <stdio.h>
#include <stdlib.h>

typedef struct Celula {
    int elemento;
    struct Celula* prox;
} Celula;

Celula* ConstrutorCelula(int x) {
    Celula* novaCelula = (Celula*) malloc(sizeof(Celula));
    if (novaCelula != NULL) {
        novaCelula->elemento = x;
        novaCelula->prox = NULL;
    }
    return novaCelula;
}

typedef struct Pilha {
    Celula* topo;
} Pilha;

Pilha ConstrutorPilha() {
    Pilha pilha;
    pilha.topo = NULL;
    return pilha;
}

void inserir(Pilha* pilha, int x) {
    Celula* novaCelula = ConstrutorCelula(x);
    if (novaCelula != NULL) {
        novaCelula->prox = pilha->topo;
        pilha->topo = novaCelula;
    }
}

int remover(Pilha* pilha){
    int elemento = -1;
    if(pilha->topo == NULL){
        printf("PILHA VAZIA");
    }else{
        Celula* tmp = pilha->topo;
        pilha->topo = pilha->topo->prox;

        elemento = tmp->elemento;
        tmp->prox = NULL;
    }

    return elemento;
}

void mostra(Pilha pilha) {
    for (Celula* tmp = pilha.topo; tmp != NULL; tmp = tmp->prox) {
        printf("%d ", tmp->elemento);
    }
    printf("\n");
}

int main() {
    Pilha pilha = ConstrutorPilha();
    inserir(&pilha, 1);
    mostra(pilha);
    inserir(&pilha, 2);
    mostra(pilha);
    inserir(&pilha, 3);
    mostra(pilha);
    remover(&pilha);
    mostra(pilha);


    while (pilha.topo != NULL) {
        Celula* tmp = pilha.topo;
        pilha.topo = pilha.topo->prox;
        free(tmp);
    }

    return 0;
}
